doc/neps/nep-0034.rst

```diff
@@ -0,0 +1,118 @@
+raise a ``ValueError`` instead.
+
+Motivation and Scope
+--------------------
+
+Users who specify lists-of-lists when creating a `numpy.ndarray` via
+``np.array`` may mistakenly pass in lists of different lengths. Currently we
+accept this input and create a ragged array with ``dtype=object``. This can be
+confusing, since it is rarely what is desired. Changing the automatic dtype
+detection to never return ``object`` for ragged arrays (defined as a recursive
+`sequence`_ of sequences, where not all the sequences on the same level have
+the same length) will force users who actually wish to create ``object`` arrays
+to specify that explicitly. Note that lists, tuples, and nd.arrays are all
+sequences. See for instance `issue 5303`_.
+
+Usage and Impact
+----------------
+
+After this change, ragged array creation must explicitly define a dtype:
+
+    >>> np.array([[1, 2], [1]])
+    ValueError: cannot guess the desired dtype from the input
+
+    >>> np.array([[1, 2], [1]], dtype=object)
+    # succeeds, with no change from current behaviour
+
+Detailed description
+--------------------
+
+To explicitly set the shape of the object array, since it is sometimes hard to
+determine what shape is desired, one could use:
+
+    >>> arr = np.empty(correct_shape, dtype=object)
+    >>> arr[...] = values
+
+We will also reject mixed seqeunces of ``non-sequence and sequence``, for instance
+all of these will be rejected:
+
+    >>> arr = np.array([np.arange(10), [10]])
+    >>> arr = np.array([[range(3), range(3), range(3)], [range(3), 0, 0]])
+
+Related Work
+------------
+
+`PR 14341`_ tried to raise an error when ragged arrays were specified with
+a numeric dtype ``np.array, [[1], [2, 3]], dtype=int)`` but failed due to
+false-positives, for instance ``np.array([1, np.array([5])], dtype=int)``.
+
+.. _`PR 14341`: https://github.com/numpy/numpy/pull/14341
+
+Implementation
+--------------
+
+The code to be changed is inside ``PyArray_DTypeFromObject``, specifically in
+``PyArray_DTypeFromObjectHelper``. Since ``PyArray_DTypeFromObject`` is part of
+the NumPy C-API, its interface cannot be changed, but it can return ``-1`` to
+indicate failure.
+
+Backward compatibility
+----------------------
+
+Anyone depending on ragged lists-of-lists creating object arrays will need to
+modify their code. There will be a deprecation period during which the current
+behaviour will emit a ``DeprecationWarning``.
+
+
+Alternatives
+------------
+
+We could continue with the current situation.
+
+It was also suggested to add a kwarg `depth` to array creation, or perhaps to
+add another array creation API function `ragged_array_object`. The goal was
+to eliminate the ambiguity in creating an object array from `array([[1, 2],
+[1]], dtype=object)`: should the returned array have a shape of `(1,)`, or
+`(2,)`? This NEP does not deal with that issue, and only deprecates the use of
+`array` with no `dtype=object` for ragged arrays.
```

Comment:
Double ``` `` ``` throughout

Reply:
Perhaps worth adding "as a consequence of choosing not to deal with this issue, users of ragged arrays may be faced with a second deprecation cycle in the future" or something.

I think I agree with declaring this out of scope - there's a lot of value to making noise when ragged arrays weren't intended.

Reply:
> there's a lot of value to making noise when ragged arrays weren't intended.

I agree

> I think I agree with declaring this out of scope 

I don't mind either, but I would suggest to do whatever is easier (I don't know what that is). If just ragged arrays are easier, perhaps give that as a rationale. If we have to jump through extra hoops to do _just_ ragged arrays (IIRC we failed once before?), then why bother?